ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineContext.java

```diff
@@ -208,16 +238,71 @@ DdlCommand createDdlCommand(
   String executeDdl(
       final String sqlExpression,
       final DdlCommand command,
-      final boolean withQuery
+      final boolean withQuery,
+      final Set<SourceName> withQuerySources
   ) {
-    final DdlCommandResult result = ddlCommandExec.execute(sqlExpression, command, withQuery);
+    if (command instanceof DropSourceCommand) {
+      throwIfOtherQueriesExist(((DropSourceCommand) command).getSourceName());
+    }
+
+    final DdlCommandResult result =
+        ddlCommandExec.execute(sqlExpression, command, withQuery, withQuerySources);
     if (!result.isSuccess()) {
       throw new KsqlStatementException(result.getMessage(), sqlExpression);
     }
+
+    if (command instanceof DropSourceCommand) {
+      // terminate the query (linked by create_as commands) after deleting the source to avoid
+      // other commands to create queries from this source while the query is being terminated
+      maybeTerminateCreateAsQuery(((DropSourceCommand) command).getSourceName());
+    }
+
     return result.getMessage();
   }
 
-  void registerQuery(final QueryMetadata query) {
+  private void maybeTerminateCreateAsQuery(final SourceName sourceName) {
+    createAsQueries.computeIfPresent(sourceName, (ignore , queryId) -> {
+      persistentQueries.get(queryId).close();
+      return null;
+    });
+  }
+
+  private Set<QueryId> getOtherQueries(
+      final SourceName sourceName,
+      final BiPredicate<SourceName, PersistentQueryMetadata> filterQueries
+  ) {
+    return otherQueries.getOrDefault(sourceName, Collections.emptySet()).stream()
+        .map(persistentQueries::get)
+        .filter(query -> filterQueries.test(sourceName, query))
+        .map(QueryMetadata::getQueryId)
+        .collect(Collectors.toSet());
+  }
+
+  private void throwIfOtherQueriesExist(final SourceName sourceName) {
+    final Set<QueryId> sinkQueries = getOtherQueries(sourceName, FILTER_QUERIES_WITH_SINK);
+    final Set<QueryId> sourceQueries = getOtherQueries(sourceName, FILTER_QUERIES_WITH_SOURCE);
```

Comment:
I know that this part is enforced by the metastore, which is why we don't use `FILTER_QUERIES_WITH_SOURCE` anywhere in this class other than here. That seems a little fragile to rely on this lose coupling. I think it's OK for now, but at a minimum add a comment so the next person debugging knows where to look